### PR TITLE
Introduce a `search_throttled` threadpool

### PR DESCRIPTION
**Pull request by s1monw**
Today all searches happen on the search threadpool which is the correct
behavior in almost any case. Yet, there are exceptions where for instance
searches searches should be passed through a single-thread
threadpool to reduce impact on a node. This change adds a index-private setting that allows to mark an index as throttled for searches and forks off all non-stats searcher access to this threadpool for indices that are marked as `index.search.throttled`

### CODE REVIEW

**Comment by s1monw** [comment]
the majority of the change is cutting SearchService over to async execution. The actual addition of the setting and such is really a smallish change.

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 1]
```diff
@@ -405,59 +423,64 @@ private QueryFetchSearchResult executeFetchPhase(SearchContext context, SearchOp
...
-        } catch (Exception e) {
-            operationListener.onFailedQueryPhase(context);
-            logger.trace("Query phase failed", e);
-            processFailure(context, e);
-            throw ExceptionsHelper.convertToRuntime(e);
-        } finally {
-            cleanContext(context);
-        }
+    public void executeQueryPhase(InternalScrollSearchRequest request, SearchTask task,
+                                                     ActionListener<ScrollQuerySearchResult> listener) {
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/search/SearchService.java]: Nit: ident

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 2]
```diff
@@ -1050,4 +1106,32 @@ public IndicesService getIndicesService() {
...
+        private boolean canMatch;
+
+        public CanMatchResponse() {
+        }
+
+        public CanMatchResponse(boolean canMatch) {
+            this.canMatch = canMatch;
+        }
+
+
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/search/SearchService.java]: Nit: extra line

**Review by tlrx — APPROVED** [review]
LGTM - I didn't expect this change to be so simple.

**Comment by jpountz** [comment]
I'm curious why we now fork can_match tasks unless the default threadpool is used, what is the benefit compared to never forking?

File: server/src/main/java/org/elasticsearch/index/IndexSettings.java [hunk 3]
```diff
@@ -277,6 +278,26 @@
...
+     * or search and generic. Other build-in threadpools are disallowed.
+     */
+    public static final Setting<String> INDEX_SEARCH_THREAD_POOL =
+        new Setting<>("index.search.threadpool", ThreadPool.Names.SEARCH, s -> {
+            if (s == null || s.isEmpty()) {
+                throw new IllegalArgumentException("Value for [index.search.threadpool] must be a non-empty string.");
+            }
+            if (ThreadPool.Names.SEARCH.equals(s) == false && ThreadPool.Names.GENERIC.equals(s) == false &&
+                ThreadPool.THREAD_POOL_TYPES.containsKey(s)) {
+                throw new IllegalArgumentException("Invalid valid for [index.search.threadpool] - " + s + " is a reserved built-in " +
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/index/IndexSettings.java]: `valid` -> `value`

**Review by jasontedor — CHANGES_REQUESTED** [review]
I have some concerns.

**Comment by s1monw** [comment]
@jasontedor here is my suggestion. 
 1. I will add a `search_single` threadpool with a single thread and a queue.
 2. users can either decide to go to `search_single` or `search`
 3. in a seperate PR I add an option on the SearchRequest that does a _force put_ on the queue that way we can fix #33205 by special casing where necessary and not globally on the index.

@jpountz with 2. I think your question is more clear as well. for the  `search_single` case we need to make sure nobody accesses more than one reader of an index that is marked as `search_single` at the same time. 

I guess that makes this change more contained and addresses all concerns.

**Comment by jasontedor** [comment]
@s1monw The plan sounds good, I think that we are mostly on the same page. I do wonder if we need to allow users to use `search_single` AKA `frozen`, i think we can do it automatically for frozen indices and otherwise use `search`? Then this would not need to be user configurable at all?

**Comment by s1monw** [comment]
> @s1monw The plan sounds good, I think that we are mostly on the same page. I do wonder if we need to allow users to use search_single AKA frozen, i think we can do it automatically for frozen indices and otherwise use search? Then this would not need to be user configurable at all?

I can do that. Private setting sounds good to me. then this PR will only add the setting and the reading. The test can add a plugin that allows changing it.

**Comment by jasontedor** [comment]
Thanks @s1monw.

**Comment by jpountz** [comment]
The plan sounds good to me too. Then maybe call the setting "frozen_search" rather than "search_single" since it would be dedicated to frozen indices?

**Comment by s1monw** [comment]
@jasontedor @jpountz I pushed more changes

**Review by jpountz — APPROVED** [review]

**Review by jasontedor — APPROVED** [review]
Looks phenomenal @s1monw. Thank you for the extra iterations.

**Pull request merged**